V 1.1.2

dmansdman opened the pull request:
Version 1.1.2 Changelog

New commands introduced:
- listall: lists all users who have given role
- leaderboard: provides leaderboard of a given category
- leaderboardsync: one-time command to sync all users who have logged for a guild, used for leaderboard

Command names changed:
- finduser -> find
- getstats -> stats
- addorchangename -> editname
- parsevc -> parse
- unsectionsuspend -> sectionunsuspend

Command enhancements:
- ARL can use yoink and clean
- Securities can use logparse
- logkey now separates exalt keys (including inc, vial, runes. etc)

Major enhancements:
- Elite loc introduced: can configure a section to output location during an afk check
- AFK panels are left behind (like headcounts)

Minor enhancements
- Added Expose as a possible reaction
- Early reactions are slightly cleaned up
- AFK control panel slightly cleaned up
- Cleaned up command names and descriptions

Bugfixes
- Quota no longer splits between moderation roles (like helper and security)
- Helper perms added to raid instances

ewang2002 commented on src/commands/moderator/LoggingSync.ts:
```diff
@@ -0,0 +1,104 @@
+            embeds: [embed]
+        });
+
+        const allMembers = await guild.members.fetch()
```
Are you sure we want to fetch every member like this?

dmansdman replied:
Yes, this is only run a single time per guild.

ewang2002 commented on src/commands/moderator/LoggingSync.ts:
```diff
@@ -0,0 +1,104 @@
+            cmdCode: "LOGGING_SYNC_COMMAND",
+            formalCommandName: "Logging Sync",
+            botCommandName: "loggingsync",
+            description: "Updates list of users who have logged items for this guild.",
```
What is this command doing, exactly?

dmansdman replied:
Updating each guild's usersWithLogs to capture all the data up to this point. Future entries are automatically populated by the logging manager.

ewang2002 commented on src/constants/dungeons/MappedAfkCheckReactions.ts:
```diff
@@ -8,7 +8,8 @@ export const MAPPED_AFK_CHECK_REACTIONS: IMappedAfkCheckReactions = {
             identifier: EmojiConstants.NITRO_EMOJI
         },
-        name: "Nitro"
+        name: "Nitro",
```
I bet this was fun to do, right?

ewang2002 commented on src/instances/RaidInstance.ts:
```diff
@@ -1523,6 +1533,7 @@ export class RaidInstance {
                 .append(`the raid location. Your new location is: **${this._location}**.`)
                 .toString()
         });
+        this.sendLocToElite();
```
This should be awaited, no? Or else you can throw a `.then()` after the call

dmansdman replied:
No need to await, nothing depends on this method completing.  can add a then().

ewang2002 replied:
I see. Usually in said case I just stick a `.then()` so my IDE doesn't give me a warning. But not a big deal as you said

ewang2002 commented on src/managers/MongoManager.ts:
```diff
@@ -523,6 +524,7 @@ export namespace MongoManager {
             moderation: {blacklistedUsers: [], suspendedUsers: [], blacklistedModmailUsers: [], mutedUsers: []},
             otherMajorConfig: getOtherMajorConfigObj(),
             properties: {
+                usersWithLogs: [],
```
Remind me what this does, again?

dmansdman replied:
For the leaderboard, I don't want to search through EVERY user on a guild, just those who have logged runs/keys/etc with the guild.  This keeps track of those people, keeps the search time down.  This is updated through either running loggingsync, which does check every user, and through the logging manager, which runs each time a log is ran.

ewang2002 commented on src/utilities/ArrayUtilities.ts:
```diff
@@ -156,6 +156,10 @@ export namespace ArrayUtilities {
      * @return {T[][]} The array of subsets of array.
      */
     export function breakArrayIntoSubsets<T>(arr: T[], size: number): T[][] {
+        if(!arr){
```
Why do we need this? The parameter `T[]` should guarantee some array, not `null` or `undefined` or anything like that.

dmansdman replied:
Yeah it was a bad attempt to fix it crashing on an empty arr, I worked around it later but forgot to remove this.  Will clean up.

dmansdman replied:
that being said, I recommend revisiting this function so it doesnt lead to crashes on certain input.

ewang2002 replied:
Under what condition would this function lead to an error?

ewang2002 commented on src/commands/staff/Leaderboard.ts:
```diff
@@ -0,0 +1,306 @@
+import {
```
Does this command just generate a static leaderboard (i.e. a leaderboard embed that doesn't change, even after new leaders are added or existing leaders add more runs)?

If this is the case, and you expect the public to see this, you might consider having the bot send a separate message containing the embed (right now, the bot just edits its interaction, which means people can see the command that was executed).

dmansdman replied:
This generates the leaderboard each time the command is called.  Like Ooga.

dmansdman replied:
Deciding against this, since it adds complexity to the dungeon selection panel.  This is the same way that getstats and finduser are implemented so I don't see a reason to change it.

ewang2002 replied:
Oh I see what you meant. I thought this was displaying the leaderboard to everyone (like in a leaderboard channel). Makes sense.

dmansdman commented:
If you want to hop in a vc and discuss the leaderboard implementation, just let me know.  I can cover how usersWithLogs is used in relation to the leaderboard, and the reason behind the loggingsync command.

dmansdman commented:
Adding expose as a reaction

dmansdman commented:
Cleaned up command names and descriptions.  Also renamed loggingsync to leaderboardsync for clarification.  Check changelog for all command name changes

ewang2002 commented:
I think this looks fine. I will approve, but not merge, this PR. 

When you think everything looks fine, feel free to merge (or else let me know in case you need a review again).

dmansdman commented:
Fine to merge, merging
